pkg/util/chunk/chunk_in_disk.go

```diff
@@ -0,0 +1,320 @@
+	diskTracker   *disk.Tracker // track disk usage.
+
+	dataFile diskFileReaderWriter
+
+	// Write or read data needs this buffer to temporarily store data
+	buf []byte
+}
+
+// NewDataInDiskByChunks creates a new DataInDiskByChunks with field types.
+func NewDataInDiskByChunks(fieldTypes []*types.FieldType) *DataInDiskByChunks {
+	l := &DataInDiskByChunks{
+		fieldTypes:    fieldTypes,
+		totalDataSize: 0,
+		totalRowNum:   0,
+		// TODO: set the quota of disk usage.
+		diskTracker: disk.NewTracker(memory.LabelForChunkDataInDiskByChunks, -1),
+		buf:         make([]byte, 0, 4096),
+	}
+	return l
+}
+
+func (d *DataInDiskByChunks) initDiskFile() (err error) {
+	err = disk.CheckAndInitTempDir()
+	if err != nil {
+		return
+	}
+	err = d.dataFile.initWithFileName(defaultChunkDataInDiskByRowsPath + strconv.Itoa(d.diskTracker.Label()))
+	if err != nil {
+		return
+	}
+	return
+}
+
+// GetDiskTracker returns the memory tracker of this List.
+func (d *DataInDiskByChunks) GetDiskTracker() *disk.Tracker {
+	return d.diskTracker
+}
+
+// Add adds a chunk to the DataInDiskByChunks. Caller must make sure the input chk has the same field types.
+// Warning: Do not concurrently call this function.
+func (d *DataInDiskByChunks) Add(chk *Chunk) (err error) {
+	if chk.NumRows() == 0 {
+		return errors2.New("Chunk spilled to disk should have at least 1 row")
+	}
+
+	if d.dataFile.file == nil {
+		err = d.initDiskFile()
+		if err != nil {
+			return
+		}
+	}
+
+	serializedBytesNum := d.serializeDataToBuf(chk)
+
+	var writeNum int
+	writeNum, err = d.dataFile.write(d.buf)
+	if err != nil {
+		return
+	}
+
+	if int64(writeNum) != serializedBytesNum {
+		return errors2.New("Some data fail to be spilled to disk")
+	}
+	d.offsetOfEachChunk = append(d.offsetOfEachChunk, d.totalDataSize)
+	d.totalDataSize += serializedBytesNum
+	d.totalRowNum += int64(chk.NumRows())
+	d.dataFile.offWrite += serializedBytesNum
+
+	d.diskTracker.Consume(serializedBytesNum)
+	return
+}
+
+func (d *DataInDiskByChunks) getChunkSize(chkIdx int) int64 {
+	totalChunkNum := len(d.offsetOfEachChunk)
+	if chkIdx == totalChunkNum-1 {
+		return d.totalDataSize - d.offsetOfEachChunk[chkIdx]
+	}
+	return d.offsetOfEachChunk[chkIdx+1] - d.offsetOfEachChunk[chkIdx]
+}
+
+// GetChunk gets a Chunk from the DataInDiskByChunks by chkIdx.
+func (d *DataInDiskByChunks) GetChunk(chkIdx int) (*Chunk, error) {
+	reader := bufio.NewReader(d.dataFile.getSectionReader(d.offsetOfEachChunk[chkIdx]))
+	chkSize := d.getChunkSize(chkIdx)
+
+	if cap(d.buf) < int(chkSize) {
+		d.buf = make([]byte, chkSize)
+	} else {
+		d.buf = d.buf[:chkSize]
+	}
+
+	readByteNum, err := io.ReadFull(reader, d.buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if int64(readByteNum) != chkSize {
+		return nil, errors2.New("Fail to restore the spilled chunk")
+	}
+
+	chk := NewEmptyChunk(d.fieldTypes)
+	d.deserializeDataToChunk(chk)
+
+	return chk, nil
+}
+
+// Close releases the disk resource.
+func (d *DataInDiskByChunks) Close() error {
```

Comment:
So it’s impossible to return err here, right?
```suggestion
func (d *DataInDiskByChunks) Close() {
```

Reply:
> So it’s impossible to return err here, right?

done